### PR TITLE
[mirai] cleanup mirai across storage/types/mempool

### PR DESCRIPTION
**Pull request by lightmark**
### Description

Per @alinush 's request. clean mirai up.

### Test Plan
ut

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/3533)

### CODE REVIEW

**Review by msmouse — APPROVED** [review]

File: storage/accumulator/src/lib.rs [hunk 1]
```diff
@@ -318,9 +317,6 @@ where
     ///     and the full route from root of that subtree to the accumulator root turns frozen
     ///         height - (log2(num_new_leaves) + 1) < height - 1 = root_level
     fn max_to_freeze(num_new_leaves: usize, root_level: u32) -> usize {
-        precondition!(root_level as usize <= MAX_ACCUMULATOR_PROOF_DEPTH);
```
Comment (alinush) [inline storage/accumulator/src/lib.rs:321]: ...but add `debug_asserts!` !!!!!

Comment (alinush) [reply storage/accumulator/src/lib.rs:321]: No?

File: storage/jellyfish-merkle/src/lib.rs [hunk 2]
```diff
@@ -960,7 +960,6 @@ trait NibbleExt {
 impl NibbleExt for HashValue {
     /// Returns the `index`-th nibble.
     fn get_nibble(&self, index: usize) -> Nibble {
-        mirai_annotations::precondition!(index < HashValue::LENGTH);
```
Comment (alinush) [inline storage/jellyfish-merkle/src/lib.rs:963]: `debug_assert!`?

File: storage/jellyfish-merkle/src/restore/mod.rs [hunk 3]
```diff
@@ -89,7 +88,6 @@ where
     }
 
     fn set_child(&mut self, index: usize, child_info: ChildInfo<K>) {
-        precondition!(index < 16);
```
Comment (alinush) [inline storage/jellyfish-merkle/src/restore/mod.rs:92]: `debug_assert!`?

File: types/src/nibble/nibble_path/mod.rs [hunk 4]
```diff
@@ -91,14 +90,12 @@ prop_compose! {
 impl NibblePath {
     /// Creates a new `NibblePath` from a vector of bytes assuming each byte has 2 nibbles.
     pub fn new_even(bytes: Vec<u8>) -> Self {
-        checked_precondition!(bytes.len() <= ROOT_NIBBLE_HEIGHT / 2);
```
Comment (alinush) [inline types/src/nibble/nibble_path/mod.rs:94]: `debug_assert!`?

File: types/src/nibble/nibble_path/mod.rs [hunk 5]
```diff
@@ -91,14 +90,12 @@ prop_compose! {
...
         let num_nibbles = bytes.len() * 2;
         NibblePath { num_nibbles, bytes }
     }
 
     /// Similar to `new()` but assumes that the bytes have one less nibble.
     pub fn new_odd(bytes: Vec<u8>) -> Self {
-        checked_precondition!(bytes.len() <= ROOT_NIBBLE_HEIGHT / 2);
```
Comment (alinush) [inline types/src/nibble/nibble_path/mod.rs:101]: `debug_assert!`?

File: types/src/nibble/nibble_path/mod.rs [hunk 6]
```diff
@@ -113,18 +110,15 @@ impl NibblePath {
     }
 
     fn new_from_byte_array(bytes: &[u8], num_nibbles: usize) -> Self {
```
Comment (alinush) [inline types/src/nibble/nibble_path/mod.rs:112]: `debug_assert!`'s ?

**Review by alinush — DISMISSED** [review]
Shouldn't we replace every MIRAI assumption/check with `debug_assert!`?

File: storage/accumulator/src/lib.rs (hunk 1 above)
Comment (msmouse) [reply storage/accumulator/src/lib.rs:321]: They didn't justify a runtime assert back then (probably more here just for mirai people having fun). I tend to agree we can just remove them. Probably have a second look case by case, but definitely not replacing every one of them an assertion.

Comment (alinush) [reply storage/accumulator/src/lib.rs:321]: Not a runtime assertion, but why not a debug assertion?

File: storage/jellyfish-merkle/src/restore/mod.rs (hunk 3 above)
Comment (msmouse) [reply storage/jellyfish-merkle/src/restore/mod.rs:92]: Too verbose.

File: storage/accumulator/src/lib.rs (hunk 1 above)
Comment (msmouse) [reply storage/accumulator/src/lib.rs:321]: Too verbose? 

Also it's like overflow-check we were discussing, if we check it in debug build, we should check it in release.

**Comment by lightmark** [comment]
> Unlike [assert!](https://doc.rust-lang.org/std/macro.assert.html), debug_assert! statements are only enabled in non optimized builds by default. An optimized build will not execute debug_assert! statements unless -C debug-assertions is passed to the compiler. This makes debug_assert! useful for checks that are too expensive to be present in a release build but may be helpful during development.

These checks are not too expensive.. so either we assert them or remove them.

File: storage/accumulator/src/lib.rs (hunk 1 above)
Comment (alinush) [reply storage/accumulator/src/lib.rs:321]: But, AFAICT, this is not like the overflow check at all: this is just a good sanity check for mis-uses of this function. When we are debugging an issue in debug mode, this assert might save you a lot of trouble identifying the culprit.

Comment (lightmark) [reply storage/accumulator/src/lib.rs:321]: debug_assertion is for too expensive checking in release mode. If it is important to check and not expensive, we should use `assert!` instead.

Comment (msmouse) [reply storage/accumulator/src/lib.rs:321]: That's why we can go case eny case: to identify if it's worth nothing, a comment or an assertion. 

But to move faster I'm okay with making them all assertions as well.

**Comment by msmouse** [comment]
Agreed, a debug assertion can give you the illusion that someyhing is impossible but it happens in product. I'd rather do nothing or a real assertion.

**Review by davidiw — APPROVED** [review]

File: storage/accumulator/src/lib.rs (hunk 1 above)
Comment (alinush) [reply storage/accumulator/src/lib.rs:321]: I agree with @lightmark that debug asserts are useful for avoiding expensive checks in release mode.

I also see @davidiw's point about the inconsistencies between dev and prod...

Not sure what the right path is here, so deferring to you.

Comment (davidiw) [reply storage/accumulator/src/lib.rs:321]: What I'd like to see is some solid thinking around what should and shouldn't be feasible and then some fuzzing that ensures that those expectations are met.

Comment (alinush) [reply storage/accumulator/src/lib.rs:321]: Isn't the first half of that called formal verification? 😛

**Pull request merged**